### PR TITLE
General performance improvements

### PR DESCRIPTION
**Pull request by freya022**
Getting an emoji by its alias is now a *lot* faster, mainly due to avoiding an O(N) loop over all the emojis in the worst case (i.e., when the alias is invalid)

Granted this is only a 5 second warmup/iteration, on 5 warmups/iterations, on one fork:

Old code:
image

New code:
image

The alias maps are lazily created, as I believe most people would either not use it, or stick to one alias group.

## Duplicate refactoring
[Most of the code](https://github.com/felldo/JEmoji/blob/badbc4e2d885511de308ea25aca4d109c554a715/lib/src/main/java/net/fellbaum/jemoji/EmojiManager.java#L380-L413) finding unicode emojis is duplicated in each method, each with small behavior differences, I moved the parsing logic away in a single method, calling an interface on certain points

## Future improvements
Some parts of the library use `List#contains` when using `Set#contains` would be faster in most cases, I saw that Java 8 does not provide an easy, nor performant, way to transform an array into a `Set`.

Since Java 8 is already EOL since March 2022, I would strongly consider updating JEmoji to Java 17 (which also aligns with Javacord)

This would enable usages of `Set.of` (replacing `Arrays.asList`) and `Set.copyOf` (replacing constructs such as `Collections.unmodifiableList(new ArrayList<>(aliases))`)

This would also change `List<Alias>` into `Set<Alias>` in `Emoji`

### CODE REVIEW

File: lib/src/main/java/net/fellbaum/jemoji/EmojiManager.java [hunk 1]
```diff
@@ -474,22 +470,63 @@ public static String replaceEmojis(final String text, final String replacementSt
...
+     * @return The text with all emojis replaced.
+     */
+    public static String replaceEmojis(final String text, Function<Emoji, String> replacementFunction, final Emoji... emojisToReplace) {
+        return replaceEmojis(text, replacementFunction, Arrays.asList(emojisToReplace));
+    }
+
+    private static <T> T forEachEmoji(final String text,
+                                      final EmojiProcessor<T> processor) {
+        final int[] textCodePointsArray = stringToCodePoints(text);
+        final long textCodePointsLength = textCodePointsArray.length;
```
Comment (freya022) [inline lib/src/main/java/net/fellbaum/jemoji/EmojiManager.java:517]: This is a hot spot, there's a way to get the code point from the string, which would save us from getting the entire array (except in worst case of course), but i couldn't figure out a way to get the correct index

Might be interesting though: https://stackoverflow.com/a/1527891 and `String#offsetByCodePoints`

**Comment by felldo** [comment]
I really appreciate your work on this. Do you mind creating separate PR's for every change you do? 
For example the jitpack changes, code refactoring (maybe even more, haven't really had a look on everything yet) and from your other [PR](https://github.com/felldo/JEmoji/pull/12) changing the project name can all be their own PR.
This would allow me to individually merge them and even more important, test their perofrmance effect on this library one by one without other changes maybe blurrying the impact of other changes.

**Comment by felldo** [comment]
> Since Java 8 is already EOL since March 2022, I would strongly consider updating JEmoji to Java 17

I'm all in for upgrading, but the corporate world is how it is and I want to offer a replacement for the many unmaintained libraries. The ideal solution is to get the library into a solid state, and then if possible swap the EmojiManager (more recent java version) with the old java 8 version if they are running on a lower version that does not support the current version. The file would not get any updates but it still exists if someone is still running on 8. And there are still quite some applications that are [running on 8](https://newrelic.com/resources/report/2023-state-of-the-java-ecosystem)

**Comment by freya022** [comment]
Replacing with smaller PRs

**Pull request closed without merging**